## Preface

Due to easy availibilty of enormous items(services) on your favourite online platforms like *e-commerce*, *job portal*, *food delivery*, *music or video streaming*, it is very hard and time consuming to find out the desired item of your choice quickly. These platforms could help you by **recommending** items as per your interest and preference by just analyzing your past interaction or behaviour with the system.  

From **Amazon** to **Linkedin**, **Uber eats** to **Spotify**, **Netflix** to **Facebook**, **Recommender systems** are most extensively used to suggest "Similar items", "Relevant jobs", "preferred foods", "Movies of interest" etc to their users. 

**Recommender system** with appropiate item suggestions helps in boosting sales, increasing revenue, retaining customers and also adds competitive advantage. 
There are basically two kind of **recommendation** methods.
1. **Content based recommendation**
2. **Collaborative filtering**

**Content based recommendation ** is based on similarity among users/items obtained through their **attributes**. It uses the additional information(meta data) about the **users** or **items** i.e. it relies on what kind of **content** is already available. This meta data could be **user's demograpic information** like *age*, *gender*, *job*, *location*, *skillsets* etc. Similarly for **items** it can be *item name*, *specifications*, *category*, *registration date* etc.

So the core idea is to recommend items by finding similar items/users to the concerned **item/user** based on their **attributes**. 

In this kernel, I am going to discuss about **Content based recommendation** using **News category** dataset. The goal is to recommend **news articles** which are similar to the already read article by using attributes like article *headline*, *category*, *author* and *publishing date*.

So let's get started without any further delay.

## Notebook - Table of Content

1. [**Importing necessary Libraries**](# 1.-Importing-necessary-Libraries)   
2. [**Loading Data**](#2.-Loading-Data)  
3. [**Data Preprocessing**](#3.-Data-Preprocessing)  
    3.a [**Fetching only the articles from 2018**](#3.a-Fetching-only-the-articles-from-2018)  
    3.b [**Removing all the short headline articles**](#3.b-Removing-all-the-short-headline-articles)  
    3.c [**Checking and removing all the duplicates**](#3.c-Checking-and-removing-all-the-duplicates)  
    3.d [**Checking for missing values**](#3.d-Checking-for-missing-values)  
4. [**Basic Data Exploration**](#4.-Basic-Data-Exploration)  
    4.a [**Basic statistics - Number of articles,authors,categories**](#4.a-Basic-statistics---Number-of-articles,authors,categories)  
    4.b [**Distribution of articles category-wise**](#4.b-Distribution-of-articles-category-wise)  
    4.c [**Number of articles per month**](#4.c-Number-of-articles-per-month)   
    4.d [**PDF for length of headlines**](#4.d-PDF-for-length-of-headlines)
5. [**Text Preprocessing**](#5.-Text-Preprocessing)  
    5.a [**Stopwords removal**](#5.a-Stopwords-removal)  
    5.b [**Lemmatization**](#5.b-Lemmatization)  
6. [**Headline based similarity on new articles**](#6.-Headline-based-similarity-on-new-articles)  
    6.a [**Using Bag of Words method**](#6.a-Using-Bag-of-Words method)  
    6.b [**Using TF-IDF method**](#6.b-Using-TF-IDF-method)  
    6.c [**Using Word2Vec embedding**](#6.c-Using-Word2Vec-embedding)   
    6.d [**Weighted similarity based on headline and category**](#6.d-Weighted-similarity-based-on-headline-and-category)  
    6.e [**Weighted similarity based on headline, category and author**](#6.e-Weighted-similarity-based-on-headline,-category-and-author)  
    6.f [**Weighted similarity based on headline, category, author and publishing day**](#6.f-Weighted-similarity-based-on-headline,-category,-author-and-publishing-day)  
 

## 1. Importing necessary Libraries

In [1]:
import numpy as np
import pandas as pd

import os
import math
import time

import matplotlib.pyplot as plt
import seaborn as sns
import plotly.figure_factory as ff
import plotly.graph_objects as go
import plotly.express as px

# Below libraries are for text processing using NLTK
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer

# Below libraries are for feature representation using sklearn
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer

# Below libraries are for similarity matrices using sklearn
from sklearn.metrics.pairwise import cosine_similarity  
from sklearn.metrics import pairwise_distances

## 2. Loading Data

In [2]:
news_articles = pd.read_json("/kaggle/input/news-category-dataset/News_Category_Dataset_v2.json", lines = True)

In [3]:
news_articles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200853 entries, 0 to 200852
Data columns (total 6 columns):
category             200853 non-null object
headline             200853 non-null object
authors              200853 non-null object
link                 200853 non-null object
short_description    200853 non-null object
date                 200853 non-null datetime64[ns]
dtypes: datetime64[ns](1), object(5)
memory usage: 9.2+ MB


The dataset contains about two million records of six different features. 

In [4]:
news_articles.head()

category                                           headline  \
0          CRIME  There Were 2 Mass Shootings In Texas Last Week...   
1  ENTERTAINMENT  Will Smith Joins Diplo And Nicky Jam For The 2...   
2  ENTERTAINMENT    Hugh Grant Marries For The First Time At Age 57   
3  ENTERTAINMENT  Jim Carrey Blasts 'Castrato' Adam Schiff And D...   
4  ENTERTAINMENT  Julianna Margulies Uses Donald Trump Poop Bags...   

           authors                                               link  \
0  Melissa Jeltsen  https://www.huffingtonpost.com/entry/texas-ama...   
1    Andy McDonald  https://www.huffingtonpost.com/entry/will-smit...   
2       Ron Dicker  https://www.huffingtonpost.com/entry/hugh-gran...   
3       Ron Dicker  https://www.huffingtonpost.com/entry/jim-carre...   
4       Ron Dicker  https://www.huffingtonpost.com/entry/julianna-...   

                                   short_description       date  
0  She left her husband. He killed their children... 2018-05-26  
1                           Of course it has a song. 2018-05-26  
2  The actor and his longtime girlfriend Anna Ebe... 2018-05-26  
3  The actor gives Dems an ass-kicking for not fi... 2018-05-26  
4  The "Dietland" actress said using the bags is ... 2018-05-26

## 3. Data Preprocessing

### 3.a Fetching only the articles from 2018  

Since the dataset size is quite large so processing through entire dataset may consume too much time. To refrain from this, we are only considering the latest articles from the year 2018. 

In [5]:
news_articles = news_articles[news_articles['date'] >= pd.Timestamp(2018,1,1)]

In [6]:
news_articles.shape

(8583, 6)

Now, the number of news articles comes down to 8583.

### 3.b Removing all the short headline articles 

After stop words removal from headline, the articles with very short headline may become blank headline articles. So let's remove all the articles with less words(<5) in the headline.   

In [7]:
news_articles = news_articles[news_articles['headline'].apply(lambda x: len(x.split())>5)]
print("Total number of articles after removal of headlines with short title:", news_articles.shape[0])

Total number of articles after removal of headlines with short title: 8530


### 3.c Checking and removing all the duplicates

Since some articles are exactly same in headlines, so let's remove all such articles having duplicate headline appearance.

In [8]:
news_articles.sort_values('headline',inplace=True, ascending=False)
duplicated_articles_series = news_articles.duplicated('headline', keep = False)
news_articles = news_articles[~duplicated_articles_series]
print("Total number of articles after removing duplicates:", news_articles.shape[0])

Total number of articles after removing duplicates: 8485


### 3.d Checking for missing values

In [9]:
news_articles.isna().sum()

category             0
headline             0
authors              0
link                 0
short_description    0
date                 0
dtype: int64

## 4. Basic Data Exploration 

### 4.a Basic statistics - Number of articles,authors,categories

In [10]:
print("Total number of articles : ", news_articles.shape[0])
print("Total number of authors : ", news_articles["authors"].nunique())
print("Total number of unqiue categories : ", news_articles["category"].nunique())

Total number of articles :  8485
Total number of authors :  892
Total number of unqiue categories :  26


### 4.b Distribution of articles category-wise

In [11]:
fig = go.Figure([go.Bar(x=news_articles["category"].value_counts().index, y=news_articles["category"].value_counts().values)])
fig['layout'].update(title={"text" : 'Distribution of articles category-wise','y':0.9,'x':0.5,'xanchor': 'center','yanchor': 'top'}, xaxis_title="Category name",yaxis_title="Number of articles")
fig.update_layout(width=800,height=700)
fig

From the bar chart, we can observe that **politics** category has **highest** number of articles then **entertainment** and so on.  

### 4.c Number of articles per month

Let's first group the data on monthly basis using **resample()** function. 

In [12]:
news_articles_per_month = news_articles.resample('m',on = 'date')['headline'].count()
news_articles_per_month

date
2018-01-31    2065
2018-02-28    1694
2018-03-31    1778
2018-04-30    1580
2018-05-31    1368
Freq: M, Name: headline, dtype: int64

In [13]:
fig = go.Figure([go.Bar(x=news_articles_per_month.index.strftime("%b"), y=news_articles_per_month)])
fig['layout'].update(title={"text" : 'Distribution of articles month-wise','y':0.9,'x':0.5,'xanchor': 'center','yanchor': 'top'}, xaxis_title="Month",yaxis_title="Number of articles")
fig.update_layout(width=500,height=500)
fig

From the bar chart, we can observe that **January** month has **highest** number of articles then **March** and so on.  

### 4.d PDF for the length of headlines 

In [14]:
fig = ff.create_distplot([news_articles['headline'].str.len()], ["ht"],show_hist=False,show_rug=False)
fig['layout'].update(title={'text':'PDF','y':0.9,'x':0.5,'xanchor': 'center','yanchor': 'top'}, xaxis_title="Length of a headline",yaxis_title="probability")
fig.update_layout(showlegend = False,width=500,height=500)
fig

The probability distribution function of headline length is almost similar to a **Guassian distribution**, where most of the headlines are 58 to 80 words long in length. 

By Data processing in Step 2, we get a subset of original dataset which has different index labels so let's make the indices uniform ranging from 0 to total number of articles. 

In [15]:
news_articles.index = range(news_articles.shape[0])

In [16]:
# Adding a new column containing both day of the week and month, it will be required later while recommending based on day of the week and month
news_articles["day and month"] = news_articles["date"].dt.strftime("%a") + "_" + news_articles["date"].dt.strftime("%b")

Since after text preprocessing the original headlines will be modified and it doesn't make sense to recommend articles by displaying modified headlines so let's copy the dataset into some other dataset and perform text preprocessing on the later.

In [17]:
news_articles_temp = news_articles.copy()

## 5. Text Preprocessing

### 5.a Stopwords removal

Stop words are not much helpful in analyis and also their inclusion consumes much time during processing so let's remove these. 

In [18]:
stop_words = set(stopwords.words('english'))

In [19]:
for i in range(len(news_articles_temp["headline"])):
    string = ""
    for word in news_articles_temp["headline"][i].split():
        word = ("".join(e for e in word if e.isalnum()))
        word = word.lower()
        if not word in stop_words:
          string += word + " "  
    if(i%1000==0):
      print(i)           # To track number of records processed
    news_articles_temp.at[i,"headline"] = string.strip()

0
1000
2000
3000
4000
5000
6000
7000
8000


### 5.b Lemmatization

Let's find the base form(lemma) of words to consider different inflections of a word same as lemma.

In [20]:
lemmatizer = WordNetLemmatizer()

In [21]:
for i in range(len(news_articles_temp["headline"])):
    string = ""
    for w in word_tokenize(news_articles_temp["headline"][i]):
        string += lemmatizer.lemmatize(w,pos = "v") + " "
    news_articles_temp.at[i, "headline"] = string.strip()
    if(i%1000==0):
        print(i)           # To track number of records processed

0
1000
2000
3000
4000
5000
6000
7000
8000


## 6. Headline based similarity on new articles

Generally, we assess **similarity** based on **distance**. If the **distance** is minimum then high **similarity** and if it is maximum then low **similarity**.
To calculate the **distance**, we need to represent the headline as a **d-dimensional** vector. Then we can find out the **similarity** based on the **distance** between vectors.

There are multiple methods to represent a **text** as **d-dimensional** vector like **Bag of words**, **TF-IDF method**, **Word2Vec embedding** etc. Each method has its own advantages and disadvantages. 

Let's see the feature representation of headline through all the methods one by one.

### 6.a Using Bag of Words method

A **Bag of Words(BoW)** method represents the occurence of words within a **document**. Here, each headline can be considered as a **document** and set of all headlines form a **corpus**.

Using **BoW** approach, each **document** is represented by a **d-dimensional** vector, where **d** is total number of **unique words** in the corpus. The set of such unique words forms the **Vocabulary**.

In [22]:
headline_vectorizer = CountVectorizer()
headline_features   = headline_vectorizer.fit_transform(news_articles_temp['headline'])

In [23]:
headline_features.get_shape()

(8485, 11122)

The output **BoW matrix**(headline_features) is a sparse matrix.

In [24]:
pd.set_option('display.max_colwidth', -1)  # To display a very long headline completely

In [25]:
def bag_of_words_based_model(row_index, num_similar_items):
    couple_dist = pairwise_distances(headline_features,headline_features[row_index])
    indices = np.argsort(couple_dist.ravel())[0:num_similar_items]
    df = pd.DataFrame({'publish_date': news_articles['date'][indices].values,
               'headline':news_articles['headline'][indices].values,
                'Euclidean similarity with the queried article': couple_dist[indices].ravel()})
    print("="*30,"Queried article details","="*30)
    print('headline : ',news_articles['headline'][indices[0]])
    print("\n","="*25,"Recommended articles : ","="*23)
    #return df.iloc[1:,1]
    return df.iloc[1:,]

bag_of_words_based_model(133, 11) # Change the row index for any other queried article

============================== Queried article details ==============================
headline :  Woman Fired After Flipping Off Trump's Motorcade Sues Former Employer

 ========================= Recommended articles :  =======================


publish_date                                                     headline  \
1  2018-04-02    The Trump Administration Is Suing California Again            
2  2018-05-01    Texas Sues Trump Administration To End DACA                   
3  2018-03-07    Stormy Daniels Suing Trump Over Nondisclosure Agreement       
4  2018-04-28    Trump: Mueller Should Never Have Been Appointed               
5  2018-04-24    Spanish Woman Looks More Like Trump Than The Donald Himself   
6  2018-02-12    What You Should Know About Trump's Nihilist Budget            
7  2018-05-09    The Caliphate Of Trump And A Planet In Ruins                  
8  2018-03-26    Trump Ally Sues Qatar For Hacking His Email                   
9  2018-02-21    All They Will Call You Will Be Deportees                      
10 2018-04-11    Pursuing Desegregation In The Trump Era                       

    Euclidean similarity with the queried article  
1   2.828427                                       
2   3.162278                                       
3   3.162278                                       
4   3.162278                                       
5   3.162278                                       
6   3.162278                                       
7   3.162278                                       
8   3.162278                                       
9   3.162278                                       
10  3.162278

Above function recommends **10 similar** articles to the **queried**(read) article based on the headline. It accepts two arguments - index of already read artile and the total number of articles to be recommended.

Based on the **Euclidean distance** it finds out 10 nearest neighbors and recommends. 

**Disadvantages**
1. It gives very low **importance** to less frequently observed words in the corpus. Few words from the queried article like "employer", "flip", "fire" appear less frequently in the entire corpus so **BoW** method does not recommend any article whose headline contains these words. Since **trump** is commonly observed word in the corpus so it is recommending the articles with headline containing "trump".   
2. **BoW** method doesn't preserve the order of words.

To overcome the first disadvantage we use **TF-IDF** method for feature representation. 


### 6.b Using TF-IDF method

**TF-IDF** method is a weighted measure which gives more importance to less frequent words in a corpus. It assigns a weight to each term(word) in a document based on **Term frequency(TF)** and **inverse document frequency(IDF)**.

**TF(i,j)** = (# times word i appears in document j) / (# words in document j)

**IDF(i,D)** = log_e(#documents in the corpus D) / (#documents containing word i)

weight(i,j) = **TF(i,j)** x **IDF(i,D)**

So if a word occurs more number of times in a document but less number of times in all other documents then its **TF-IDF** value will be high.


In [26]:
tfidf_headline_vectorizer = TfidfVectorizer(min_df = 0)
tfidf_headline_features = tfidf_headline_vectorizer.fit_transform(news_articles_temp['headline'])

In [27]:
def tfidf_based_model(row_index, num_similar_items):
    couple_dist = pairwise_distances(tfidf_headline_features,tfidf_headline_features[row_index])
    indices = np.argsort(couple_dist.ravel())[0:num_similar_items]
    df = pd.DataFrame({'publish_date': news_articles['date'][indices].values,
               'headline':news_articles['headline'][indices].values,
                'Euclidean similarity with the queried article': couple_dist[indices].ravel()})
    print("="*30,"Queried article details","="*30)
    print('headline : ',news_articles['headline'][indices[0]])
    print("\n","="*25,"Recommended articles : ","="*23)
    
    #return df.iloc[1:,1]
    return df.iloc[1:,]
tfidf_based_model(133, 11)

============================== Queried article details ==============================
headline :  Woman Fired After Flipping Off Trump's Motorcade Sues Former Employer

 ========================= Recommended articles :  =======================


publish_date  \
1  2018-05-21     
2  2018-04-02     
3  2018-04-10     
4  2018-04-26     
5  2018-04-03     
6  2018-02-24     
7  2018-01-16     
8  2018-02-21     
9  2018-01-09     
10 2018-02-28     

                                                                        headline  \
1   The Supreme Court Just Made It A Lot Harder For You To Sue Your Employer       
2   The Trump Administration Is Suing California Again                             
3   Lou Dobbs Flips Out On Live TV, Urges Trump To 'Fire The SOB' Robert Mueller   
4   Cardi B's Former Manager Sues Her For $10 Million                              
5   A Third Woman Is Suing To Break A Trump-Related Nondisclosure Agreement        
6   Former RNC Chair Fires Back At Claim He Was Only Hired Because He Was Black    
7   State Employer Side Payroll Taxes And Loser Liberalism                         
8   Democrats Flip Kentucky State House Seat Where Trump Won Overwhelmingly        
9   Big Tax Game Hunting: Employer Side Payroll Taxes                              
10  Democrats Flip 2 More GOP-Held State House Seats                               

    Euclidean similarity with the queried article  
1   1.164067                                       
2   1.253867                                       
3   1.258810                                       
4   1.268704                                       
5   1.274264                                       
6   1.274847                                       
7   1.276696                                       
8   1.282008                                       
9   1.285147                                       
10  1.287403

Compared to **BoW** method, here **TF-IDF** method recommends the articles with headline containing words like "employer", "fire", "flip" in top 5 recommendations and these words occur less frequently in the corpus.   

**Disadvantages :- **

**Bow** and **TF-IDF** method do not capture **semantic** and **syntactic** similarity of a given word with other words but this can be captured using **Word embeddings**.

For example: there is a good association between words like "trump" and "white house", "office and employee", "tiger" and "leopard", "USA" and "Washington D.C" etc. Such kind of **semantic** similarity can be captured using **word embedding** techniques.
**Word embedding** techniques like **Word2Vec**, **GloVe** and **fastText** leverage semantic similarity between words. 

### 6.c Using Word2Vec embedding

**Word2Vec** is one of the techniques for **semantic** similarity which was invented by **Google** in 2013. For a given corpus, during training it observes the patterns and respresents each word by a **d-dimensional** vector. To get better results we need fairly large corpus.

Since our corpus size is small so let's use Google's pretrained model on **google news** articles. This standard model contains vector representation for billions of words obtained by training on millions of new articles. Here, each word is represented by a **300** dimensional dense vector. 




In [28]:
from gensim.models import Word2Vec
from gensim.models import KeyedVectors
import pickle

Since this **pre-trained Word2Vec** model is **1.5 GB** in compressed form. So it needs a high end RAM to load it in the memory after unzipping.

Here, we are loading this pre-build model from a **pickle** file which contains this model in advance.

In [29]:
os.chdir(r'/kaggle/input/')

In [30]:
!ls

googlew2v  news-category-dataset


In [31]:
with open('googlew2v/word2vec_model', 'rb') as file:
    loaded_model = pickle.load(file)

Since the model gives vector representation for each **word** but we calculate the distance between **headlines** so we need to obtain vector representation for each **headline**. One way to obtain this is by first adding vector representation of all the words available in the **headline** and then calculating the average. It is also known as **average Word2Vec** model.   

Below code cell performs the same. 

In [32]:
vocabulary = loaded_model.keys()
w2v_headline = []
for i in news_articles_temp['headline']:
    w2Vec_word = np.zeros(300, dtype="float32")
    for word in i.split():
        if word in vocabulary:
            w2Vec_word = np.add(w2Vec_word, loaded_model[word])
    w2Vec_word = np.divide(w2Vec_word, len(i.split()))
    w2v_headline.append(w2Vec_word)
w2v_headline = np.array(w2v_headline)

In [33]:
def avg_w2v_based_model(row_index, num_similar_items):
    couple_dist = pairwise_distances(w2v_headline, w2v_headline[row_index].reshape(1,-1))
    indices = np.argsort(couple_dist.ravel())[0:num_similar_items]
    df = pd.DataFrame({'publish_date': news_articles['date'][indices].values,
               'headline':news_articles['headline'][indices].values,
                'Euclidean similarity with the queried article': couple_dist[indices].ravel()})
    print("="*30,"Queried article details","="*30)
    print('headline : ',news_articles['headline'][indices[0]])
    print("\n","="*25,"Recommended articles : ","="*23)
    #return df.iloc[1:,1]
    return df.iloc[1:,]

avg_w2v_based_model(133, 11)

============================== Queried article details ==============================
headline :  Woman Fired After Flipping Off Trump's Motorcade Sues Former Employer

 ========================= Recommended articles :  =======================


publish_date  \
1  2018-03-19     
2  2018-01-26     
3  2018-02-20     
4  2018-04-03     
5  2018-02-03     
6  2018-05-04     
7  2018-03-07     
8  2018-05-03     
9  2018-02-25     
10 2018-01-25     

                                                                                       headline  \
1   White House Lawyer Insists Trump Isn't Considering Firing Mueller                             
2   White House Spent Months Denying That Trump Considered Firing Mueller                         
3   Trump Claims He 'Never Met' Woman Accusing Him Of Sexually Assaulting Her In Trump Tower      
4   17 States Sue Trump Administration Over Census Citizenship Question                           
5   Husband Of Former Trump Household Staffer Now An EPA Official                                 
6   Giuliani Tells Mueller To Back Off ‘Fine Woman’ Ivanka Trump But Calls Kushner 'Disposable'   
7   Former Trump Attorney Stuns 'Fox & Friends,' Says Stormy Daniels NDA Is Likely Invalid        
8   Giuliani Says Trump Repaid Lawyer For $130,000 Payment To Stormy Daniels                      
9   Olympian Gus Kenworthy Burns Ivanka Trump: 'TF Is She Doing Here?'                            
10  Trump HUD Official Lynne Patton Under Fire After Calling Journalist 'Miss Piggy'              

    Euclidean similarity with the queried article  
1   0.746579                                       
2   0.773220                                       
3   0.802719                                       
4   0.803982                                       
5   0.809882                                       
6   0.824002                                       
7   0.837785                                       
8   0.839581                                       
9   0.839814                                       
10  0.842145

Here, **Word2Vec** based representation recommends the headlines containing the word **white house** which is associated with the word **trump** in the queried article. Similarly, it recommends the headlines with words like "offical", "insist" which have semantic similarity to the words "employer", "sue" in the queried headline.

So far we were recommending using only one feature i.e. **headline** but in order to make a **robust** recommender system we need to consider **multiple** features at a time. Based on the business interest and rules, we can decide weight for each feature.

Let's see different models with combinations of different features for article similarity.

### 6.d Weighted similarity based on headline and category

Let's first see articles similarity based on **headline** and **category**. We are using **onehot encoding** to get feature representation for **category**.

Sometimes as per the business requirements, we may need to give more preference to the articles from the same **category**. In such cases we can assign more weight to **category** while recommending. Higher the weight, more the significance. Similarly less weight leads to less signficance to a particular feature.


In [34]:
from sklearn.preprocessing import OneHotEncoder 

In [35]:
category_onehot_encoded = OneHotEncoder().fit_transform(np.array(news_articles_temp["category"]).reshape(-1,1))

In [36]:
def avg_w2v_with_category(row_index, num_similar_items, w1,w2): #headline_preference = True, category_preference = False):
    w2v_dist  = pairwise_distances(w2v_headline, w2v_headline[row_index].reshape(1,-1))
    category_dist = pairwise_distances(category_onehot_encoded, category_onehot_encoded[row_index]) + 1
    weighted_couple_dist   = (w1 * w2v_dist +  w2 * category_dist)/float(w1 + w2)
    indices = np.argsort(weighted_couple_dist.flatten())[0:num_similar_items].tolist()
    df = pd.DataFrame({'publish_date': news_articles['date'][indices].values,
               'headline':news_articles['headline'][indices].values,
                'Weighted Euclidean similarity with the queried article': weighted_couple_dist[indices].ravel(),
                'Word2Vec based Euclidean similarity': w2v_dist[indices].ravel(),
                 'Category based Euclidean similarity': category_dist[indices].ravel(),
                'Categoty': news_articles['category'][indices].values})
    
    print("="*30,"Queried article details","="*30)
    print('headline : ',news_articles['headline'][indices[0]])
    print('Categoty : ', news_articles['category'][indices[0]])
    print("\n","="*25,"Recommended articles : ","="*23)
    #return df.iloc[1:,[1,5]]
    return df.iloc[1:, ]

avg_w2v_with_category(528,10,0.1,0.8)

============================== Queried article details ==============================
headline :  Universities Tell Applicants That Protesting Gun Violence Won’t Affect Admissions
Categoty :  EDUCATION

 ========================= Recommended articles :  =======================


publish_date  \
1 2018-02-21     
2 2018-04-02     
3 2018-02-07     
4 2018-04-16     
5 2018-04-04     
6 2018-02-06     
7 2018-04-19     
8 2018-04-06     
9 2018-02-20     

                                                                                                             headline  \
1  Texas District Says Students Protesting Gun Violence Will Get Suspended                                              
2  Teachers Swarm Kentucky Capitol To Protest Pension Changes, School Budget Cuts                                       
3  While Teachers Fight For Better Pay, West Virginia Lawmakers Discuss Opossums                                        
4  Beyoncé Announces $100,000 In Scholarships For HBCU Students                                                         
5  Oklahoma Teachers Begin 110-Mile March To Protest Education Funding                                                  
6  Homeless Students, Destroyed Campuses, ‘Invisible Injuries’: What California Schools Learned From Recent Disasters   
7  Desperate For Teachers, Districts Beg Retirees To Come Back                                                          
8  Puerto Rico To Shutter 283 More Schools This Summer As Education Crisis Deepens                                      
9  Company That Sells Bulletproof Gucci And Hermès Bags Sees Huge Sales In School Backpacks                             

   Weighted Euclidean similarity with the queried article  \
1  0.969627                                                 
2  0.988327                                                 
3  0.995150                                                 
4  0.995467                                                 
5  0.999916                                                 
6  1.002493                                                 
7  1.002621                                                 
8  1.003193                                                 
9  1.005712                                                 

   Word2Vec based Euclidean similarity  Category based Euclidean similarity  \
1  0.726643                             1.0                                   
2  0.894946                             1.0                                   
3  0.956346                             1.0                                   
4  0.959200                             1.0                                   
5  0.999242                             1.0                                   
6  1.022439                             1.0                                   
7  1.023586                             1.0                                   
8  1.028740                             1.0                                   
9  1.051409                             1.0                                   

    Categoty  
1  EDUCATION  
2  EDUCATION  
3  EDUCATION  
4  EDUCATION  
5  EDUCATION  
6  EDUCATION  
7  EDUCATION  
8  EDUCATION  
9  EDUCATION

Above function takes two extra arguments **w1** and **w2** for weights corresponding to **headline** and **category**. It is always a good practice to pass the **weights** in a range scaled from **0 to 1**, where a value close to 1 indicates high weight whereas close to 0 indicates less weight.  

Here, we can observe that the recommended articles are from the same **category** as the queried article **category**. This is due to passing of high value to **w2**.

### 6.e Weighted similarity based on headline, category and author

Now let's see calcualte articles similarity based on **author** along with **headline** and **category**. Again, we are encoding **author** through **onehot encoding**.

In [37]:
authors_onehot_encoded = OneHotEncoder().fit_transform(np.array(news_articles_temp["authors"]).reshape(-1,1))

In [38]:
def avg_w2v_with_category_and_authors(row_index, num_similar_items, w1,w2,w3): #headline_preference = True, category_preference = False):
    w2v_dist  = pairwise_distances(w2v_headline, w2v_headline[row_index].reshape(1,-1))
    category_dist = pairwise_distances(category_onehot_encoded, category_onehot_encoded[row_index]) + 1
    authors_dist = pairwise_distances(authors_onehot_encoded, authors_onehot_encoded[row_index]) + 1
    weighted_couple_dist   = (w1 * w2v_dist +  w2 * category_dist + w3 * authors_dist)/float(w1 + w2 + w3)
    indices = np.argsort(weighted_couple_dist.flatten())[0:num_similar_items].tolist()
    df = pd.DataFrame({'publish_date': news_articles['date'][indices].values,
                'headline':news_articles['headline'][indices].values,
                'Weighted Euclidean similarity with the queried article': weighted_couple_dist[indices].ravel(),
                'Word2Vec based Euclidean similarity': w2v_dist[indices].ravel(),
                'Category based Euclidean similarity': category_dist[indices].ravel(),
                'Authors based Euclidean similarity': authors_dist[indices].ravel(),       
                'Categoty': news_articles['category'][indices].values,
                'Authors': news_articles['authors'][indices].values})
    print("="*30,"Queried article details","="*30)
    print('headline : ',news_articles['headline'][indices[0]])
    print('Categoty : ', news_articles['category'][indices[0]])
    print('Authors : ', news_articles['authors'][indices[0]])
    print("\n","="*25,"Recommended articles : ","="*23)
    #return df.iloc[1:,[1,6,7]]
    return df.iloc[1:, ]


avg_w2v_with_category_and_authors(528,10,0.1,0.1,1)

============================== Queried article details ==============================
headline :  Universities Tell Applicants That Protesting Gun Violence Won’t Affect Admissions
Categoty :  EDUCATION
Authors :  Carla Herreria

 ========================= Recommended articles :  =======================


publish_date  \
1 2018-04-29     
2 2018-01-24     
3 2018-03-23     
4 2018-01-28     
5 2018-01-26     
6 2018-02-01     
7 2018-05-25     
8 2018-04-15     
9 2018-01-06     

                                                                             headline  \
1  Thousands Protest Across Spain After 5 Men Are Cleared Of Gang Rape                  
2  Trustee Defends MSU President, Dismissing Sex Abuse Reports As ‘Nassar Thing’        
3  Protests Shut Down Sacramento Kings Game, Freeways Over Stephon Clark's Death        
4  Garrison Keillor, Fired For Harassment, Goes After MPR And Accuser In Statement      
5  GOP Rep. Pat Meehan Retiring Amid Reports Of Taxpayer-Funded Harassment Settlement   
6  Rep. Adam Schiff: GOP's FBI Memo Could Lead To 'Constitutional Crisis'               
7  Rachel Dolezal Faces Felony Charges For Welfare Fraud                                
8  Prominent LGBTQ Lawyer Sets Self On Fire In 'Protest Suicide' Of Climate Change      
9  White Supremacist Charged With Terrorism After Alleged Attempt To Derail Train       

   Weighted Euclidean similarity with the queried article  \
1  1.104443                                                 
2  1.104771                                                 
3  1.106933                                                 
4  1.107877                                                 
5  1.110423                                                 
6  1.111024                                                 
7  1.111283                                                 
8  1.111568                                                 
9  1.112158                                                 

   Word2Vec based Euclidean similarity  Category based Euclidean similarity  \
1  0.839104                             2.414214                              
2  0.843041                             2.414214                              
3  0.868987                             2.414214                              
4  0.880313                             2.414214                              
5  0.910866                             2.414214                              
6  0.918069                             2.414214                              
7  0.921177                             2.414214                              
8  0.924607                             2.414214                              
9  0.931681                             2.414214                              

   Authors based Euclidean similarity      Categoty         Authors  
1  1.0                                 WORLD NEWS    Carla Herreria  
2  1.0                                 SPORTS        Carla Herreria  
3  1.0                                 BLACK VOICES  Carla Herreria  
4  1.0                                 MEDIA         Carla Herreria  
5  1.0                                 POLITICS      Carla Herreria  
6  1.0                                 POLITICS      Carla Herreria  
7  1.0                                 CRIME         Carla Herreria  
8  1.0                                 QUEER VOICES  Carla Herreria  
9  1.0                                 CRIME         Carla Herreria

Above function takes one extra weight argument **w3** for **author**.

In the ouput, we can observe that the recommended articles are from the same **author** as the queried article **author** due to high weightage to **w3**.

### 6.f Weighted similarity based on headline, category, author and publishing day 

Now let's see calcualte articles similarity based on the publishing **week day** **author** along with **headline**, **category** and **author**. Again, we are encoding this new feature through **onehot encoding**.

In [39]:
publishingday_onehot_encoded = OneHotEncoder().fit_transform(np.array(news_articles_temp["day and month"]).reshape(-1,1))

In [40]:
def avg_w2v_with_category_authors_and_publshing_day(row_index, num_similar_items, w1,w2,w3,w4): #headline_preference = True, category_preference = False):
    w2v_dist  = pairwise_distances(w2v_headline, w2v_headline[row_index].reshape(1,-1))
    category_dist = pairwise_distances(category_onehot_encoded, category_onehot_encoded[row_index]) + 1
    authors_dist = pairwise_distances(authors_onehot_encoded, authors_onehot_encoded[row_index]) + 1
    publishingday_dist = pairwise_distances(publishingday_onehot_encoded, publishingday_onehot_encoded[row_index]) + 1
    weighted_couple_dist   = (w1 * w2v_dist +  w2 * category_dist + w3 * authors_dist + w4 * publishingday_dist)/float(w1 + w2 + w3 + w4)
    indices = np.argsort(weighted_couple_dist.flatten())[0:num_similar_items].tolist()
    df = pd.DataFrame({'publish_date': news_articles['date'][indices].values,
                'headline_text':news_articles['headline'][indices].values,
                'Weighted Euclidean similarity with the queried article': weighted_couple_dist[indices].ravel(),
                'Word2Vec based Euclidean similarity': w2v_dist[indices].ravel(),
                'Category based Euclidean similarity': category_dist[indices].ravel(),
                'Authors based Euclidean similarity': authors_dist[indices].ravel(),   
                'Publishing day based Euclidean similarity': publishingday_dist[indices].ravel(), 
                'Categoty': news_articles['category'][indices].values,
                'Authors': news_articles['authors'][indices].values,
                'Day and month': news_articles['day and month'][indices].values})
    print("="*30,"Queried article details","="*30)
    print('headline : ',news_articles['headline'][indices[0]])
    print('Categoty : ', news_articles['category'][indices[0]])
    print('Authors : ', news_articles['authors'][indices[0]])
    print('Day and month : ', news_articles['day and month'][indices[0]])
    print("\n","="*25,"Recommended articles : ","="*23)
    #return df.iloc[1:,[1,7,8,9]]
    return df.iloc[1:, ]


avg_w2v_with_category_authors_and_publshing_day(528,10,0.1,0.1,0.1,1)

============================== Queried article details ==============================
headline :  Universities Tell Applicants That Protesting Gun Violence Won’t Affect Admissions
Categoty :  EDUCATION
Authors :  Carla Herreria
Day and month :  Sat_Feb

 ========================= Recommended articles :  =======================


publish_date  \
1 2018-02-24     
2 2018-02-17     
3 2018-02-17     
4 2018-02-24     
5 2018-02-03     
6 2018-02-24     
7 2018-02-24     
8 2018-02-17     
9 2018-02-03     

                                                                     headline_text  \
1  Tribal Filipinos Were A Surprising Muse For ‘Black Panther’s’ Dora Milaje         
2  Florida Gubernatorial Candidate Calls On Governor To Halt AR-15 Sales             
3  U.S. Figure Skater Nathan Chen Redeems Himself With Record-Setting Skate          
4  This Is What A Serious Gun Violence Policy Would Look Like                        
5  Uma Thurman Says Harvey Weinstein Assaulted Her                                   
6  Bombing In Syria's Ghouta Won't Let Up Long Enough For Rescuers To Count Bodies   
7  Evangelical Leaders Say 'Pro-Life Ethic' Means Fighting For Gun Reform            
8  Floridians Tell Politicians Who Do The NRA’s Bidding Their Time Is Up             
9  Bill Maher Mocks Devin Nunes’ Memo As Nothing More Than 'A Facebook Post'         

   Weighted Euclidean similarity with the queried article  \
1  1.120093                                                 
2  1.123805                                                 
3  1.131725                                                 
4  1.197894                                                 
5  1.201029                                                 
6  1.206645                                                 
7  1.206681                                                 
8  1.208333                                                 
9  1.210100                                                 

   Word2Vec based Euclidean similarity  Category based Euclidean similarity  \
1  1.146997                             2.414214                              
2  1.195254                             2.414214                              
3  1.298216                             2.414214                              
4  0.744200                             2.414214                              
5  0.784947                             2.414214                              
6  0.857960                             2.414214                              
7  0.858428                             2.414214                              
8  0.879902                             2.414214                              
9  0.902875                             2.414214                              

   Authors based Euclidean similarity  \
1  1.000000                             
2  1.000000                             
3  1.000000                             
4  2.414214                             
5  2.414214                             
6  2.414214                             
7  2.414214                             
8  2.414214                             
9  2.414214                             

   Publishing day based Euclidean similarity       Categoty  \
1  1.0                                        BLACK VOICES    
2  1.0                                        POLITICS        
3  1.0                                        SPORTS          
4  1.0                                        POLITICS        
5  1.0                                        ENTERTAINMENT   
6  1.0                                        WORLD NEWS      
7  1.0                                        POLITICS        
8  1.0                                        POLITICS        
9  1.0                                        COMEDY          

             Authors Day and month  
1  Carla Herreria     Sat_Feb       
2  Carla Herreria     Sat_Feb       
3  Carla Herreria     Sat_Feb       
4  Jonathan Cohn      Sat_Feb       
5  Hilary Hanson      Sat_Feb       
6                     Sat_Feb       
7  Carol Kuruvilla    Sat_Feb       
8  Sebastian Murdock  Sat_Feb       
9  Lee Moran          Sat_Feb

Above function takes one extra weight argument **w4** for **day of the week and month**.

In the ouput, we can observe that the recommended articles are from the same **day of the week and month** as the queried article due to high weightage to **w4**.